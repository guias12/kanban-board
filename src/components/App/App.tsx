import React from "react";
import List from "../List";
import { IList } from "../../models/types";
import { IState } from "../../reducers";
import { connect } from "react-redux";
import { DragDropContext, DropResult } from "react-beautiful-dnd";
import { draggableSort } from "../../actions";

import "./styles.scoped.scss";

interface IAppProps {
  lists: IList[];
  dispatch: any;
}

function App({ lists, dispatch }: IAppProps) {
  const handleDragEnd = (result: DropResult) => {
    const { destination, source, draggableId } = result;

    if (!destination) {
      return;
    }

    dispatch(
      draggableSort(
        source.droppableId,
        destination.droppableId,
        source.index,
        destination.index,
        draggableId
      )
    );
  };
  return (
    <DragDropContext onDragEnd={handleDragEnd}>
      <div className="App">
        <h2 className="app-title">My Kanban Board</h2>
        <div className="app-board-list">
          {lists.map((list) => {
            return (
              <List
                title={list.title}
                cards={list.cards}
                id={list.id}
                key={list.id}
              />
            );
          })}
        </div>
      </div>
    </DragDropContext>
  );
}

const mapStateToProps = (state: IState) => ({
  lists: state.lists,
});
export default connect(mapStateToProps)(App);
