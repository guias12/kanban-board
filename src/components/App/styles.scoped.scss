.App {
  background: rgb(76, 158, 8);
  background: linear-gradient(
    219deg,
    rgba(76, 158, 8, 1) 0%,
    rgba(66, 186, 25, 1) 58%,
    rgba(36, 200, 23, 1) 100%
  );
  width: 100vw;
  height: 100vh;
}

.app-title {
  color: white;
  font-size: 2em;
  margin: 0;
  padding: 15px;
}

.app-tools {
  margin: 10px 15px;
  position: relative;
}

.search-card-input {
  padding: 7px 10px;
  border-radius: 4px;
  border: none;
  width: 20%;
}

.app-board-list {
  display: flex;
  flex-direction: row;
  height: 80%;
}

@media only screen and (max-width: 600px) {
  .App {
    height: 100%;
    min-height: 100vh;
  }
  .app-board-list {
    flex-wrap: wrap;
  }
  .search-card-input {
    width: 95%;
  }
}
