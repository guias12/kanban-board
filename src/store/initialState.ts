import { IInitialState } from '../models/types';
const getInitialState = (): IInitialState => ({
  searchTerm: '',
  list: [
    {
      title: 'To Do',
      id: 0,
      cards: [
        {
          id: 0,
          text: 'Test kanban board',
        },
      ],
    },
    {
      title: 'Doing',
      id: 1,
      cards: [],
    },
    {
      title: 'Done',
      id: 2,
      cards: [],
    },
  ],
});

const getEhancedTestInitialState = (): IInitialState => ({
  searchTerm: '',
  list: [
    {
      title: 'To Do',
      id: 0,
      cards: [
        {
          id: 0,
          text: 'First ToDo card',
        },
        {
          id: 1,
          text: 'Second ToDo card',
        },
      ],
    },
    {
      title: 'Doing',
      id: 1,
      cards: [
        {
          id: 2,
          text: 'first Doing card',
        },
        {
          id: 3,
          text: 'second Doing card',
        },
        {
          id: 4,
          text: 'third Doing card',
        },
      ],
    },
    {
      title: 'Done',
      id: 2,
      cards: [
        {
          id: 5,
          text: 'first Done card',
        },
        {
          id: 6,
          text: 'second Done card',
        },
      ],
    },
  ],
});

export { getInitialState, getEhancedTestInitialState };
