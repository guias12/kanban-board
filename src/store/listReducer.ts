import { IInitialState } from '../models/types';
import { AnyAction } from 'redux';
import { CONSTANTS } from '../actions';

const initialState: IInitialState = {
  searchTerm: '',
  list: [
    {
      title: 'To Do',
      id: 0,
      cards: [
        {
          id: 0,
          text: 'Test kanban board',
        },
      ],
    },
    {
      title: 'Doing',
      id: 1,
      cards: [],
    },
    {
      title: 'Done',
      id: 2,
      cards: [],
    },
  ],
};
let nextAvailableCardId = 1;

const listReducer = (state = initialState, action: AnyAction) => {
  const stateClone: IInitialState = { ...state };

  switch (action.type) {
    case CONSTANTS.ADD_CARD:
      const newStateList = stateClone.list.map((list) => {
        if (list.id === action.payload.listId) {
          const newCardId = nextAvailableCardId;
          const newCard = {
            text: action.payload.text,
            id: newCardId,
          };
          nextAvailableCardId += 1;

          return {
            ...list,
            cards: [...list.cards, newCard],
          };
        } else {
          return list;
        }
      });

      stateClone.list = newStateList;
      state.list = newStateList;
      return stateClone;
    case CONSTANTS.CHANGE_CARD_TEXT:
      const { listId, cardId, text } = action.payload;
      let currentList = stateClone.list.find((list) => list.id === listId);
      const currentListIndex = stateClone.list.findIndex(
        (list) => list.id === listId
      );
      const edditingCardIndex = currentList?.cards.findIndex(
        (card) => card.id === cardId
      );

      if (
        (edditingCardIndex || edditingCardIndex === 0) &&
        (currentList || currentList === 0)
      ) {
        let newCards = [...currentList.cards];
        newCards[edditingCardIndex].text = text;
        currentList = {
          ...currentList,
          cards: [...newCards],
        };

        stateClone.list[currentListIndex] = currentList;
        return stateClone;
      }

      return state;
    case CONSTANTS.DRAG_CARD:
      const {
        sourceListId,
        destinationListId,
        sourceCardIndex,
        destinationCardIndex,
      } = action.payload;

      if (sourceListId === destinationListId) {
        const list = stateClone.list.find(
          (list) => sourceListId === String(list.id)
        );
        const card = list?.cards.splice(sourceCardIndex, 1);

        if (card) {
          list?.cards.splice(destinationCardIndex, 0, ...card);
        }
      } else {
        const sourceList = stateClone.list.find(
          (list) => sourceListId === String(list.id)
        );
        const card = sourceList?.cards.splice(sourceCardIndex, 1);
        const destinationList = stateClone.list.find(
          (list) => destinationListId === String(list.id)
        );
        if (card) {
          destinationList?.cards.splice(destinationCardIndex, 0, ...card);
        }
      }

      return stateClone;
    case CONSTANTS.FILTER_CARD:
      const { searchTerm } = action.payload;
      stateClone.searchTerm = searchTerm;

      return stateClone;
    default:
      return state;
  }
};

export default listReducer;
