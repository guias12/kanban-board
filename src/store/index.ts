import { configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useSelector } from 'react-redux';
import listReducer from './listSlice';

const store = configureStore({
  reducer: {
    list: listReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export default store;
