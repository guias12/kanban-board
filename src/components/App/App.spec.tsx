import React from 'react';
import { render, screen } from '@testing-library/react';
import configureMockStore, { MockStoreEnhanced } from 'redux-mock-store';
import { getInitialState } from '../../store/initialState';
import { create } from 'react-test-renderer';
import { Provider } from 'react-redux';
import App from './App';

describe('App component', () => {
  let store: MockStoreEnhanced<unknown, {}>;
  const setup = (): JSX.Element => {
    const initialState = getInitialState();
    store = configureMockStore()({ list: initialState });

    return (
      <Provider store={store}>
        <App />
      </Provider>
    );
  };
  it('Should render App component correctly', () => {
    render(setup());
    const pageTitle = screen.getByText(/My Kanban Board/i);

    expect(pageTitle).toBeInTheDocument();
  });

  it('Should match App snapshot', () => {
    const component = create(setup());
    const componentJson = component.toJSON();

    expect(componentJson).toMatchSnapshot();
  });
});
