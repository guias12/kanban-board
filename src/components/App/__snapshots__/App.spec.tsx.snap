// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Should match App snapshot 1`] = `
<div
  className="App"
>
  <h2
    className="app-title"
  >
    My Kanban Board
  </h2>
  <div
    className="app-tools"
  >
    <input
      className="search-card-input"
      id="search-cards"
      name="search-cards"
      onChange={[Function]}
      placeholder="Search card"
      type="text"
      value=""
    />
  </div>
  <div
    className="app-board-list"
  >
    <div
      className="list-container"
      data-rbd-droppable-context-id="1"
      data-rbd-droppable-id="0"
    >
      <h3>
        To Do
      </h3>
      <div
        aria-describedby="rbd-hidden-text-1-hidden-text-5"
        className="card-container"
        data-rbd-drag-handle-context-id="1"
        data-rbd-drag-handle-draggable-id="0"
        data-rbd-draggable-context-id="1"
        data-rbd-draggable-id="0"
        draggable={false}
        onDragStart={[Function]}
        onTransitionEnd={null}
        role="button"
        style={
          Object {
            "transform": null,
            "transition": null,
          }
        }
        tabIndex={0}
      >
        <span
          className="card-text"
          data-testid="card-text"
          onClick={[Function]}
        >
          Test kanban board
        </span>
      </div>
      <button
        className="list-add-card-btn"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-plus "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Create new card
      </button>
    </div>
    <div
      className="list-container"
      data-rbd-droppable-context-id="1"
      data-rbd-droppable-id="1"
    >
      <h3>
        Doing
      </h3>
      <button
        className="list-add-card-btn"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-plus "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Create new card
      </button>
    </div>
    <div
      className="list-container"
      data-rbd-droppable-context-id="1"
      data-rbd-droppable-id="2"
    >
      <h3>
        Done
      </h3>
      <button
        className="list-add-card-btn"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-plus "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Create new card
      </button>
    </div>
  </div>
</div>
`;
